### ISSUE TITLE
Optimization on SE(3)

### ISSUE THREAD

**Issue by mhubii** [facebookresearch/theseus#626]
## ❓ Questions and Help

Thank you for reading and sorry if this is a trivial question.

As an example, I am trying to find a homogenous transform such that `p_prime = HT @ p`. Please find below a script

```python
import theseus as th
import torch
import torchlie as lie

# a translation target transform
Th_target = lie.SE3.exp(torch.tensor([0.4, 0.5, 0.6, 0.0, 0.0, 0.0]))

# some point in R^3
p = torch.tensor([1.0, 0.0, 0.0]).unsqueeze(0)
p_prime = Th_target @ p

# setup optimization variables
p_var = th.Variable(p, name="p")
p_prime_var = th.Variable(p_prime, name="p_prime")
Th_vec_var = th.Vector(dof=6, name="Th_vec")


# specify cost function
def cost_fn(optim_vars, aux_vars):
    Th_vec = optim_vars[0]
    p, p_prime = aux_vars
    Th = lie.SE3.exp(Th_vec.tensor)

    p_prime_hat = Th @ p.tensor
    err = p_prime_hat - p_prime.tensor
    return err


# setup objective
objective = th.Objective()
cost_fn = th.AutoDiffCostFunction(
    optim_vars=[Th_vec_var],
    err_fn=cost_fn,
    dim=3,
    aux_vars=[p_var, p_prime_var],
    cost_weight=th.ScaleCostWeight(1.0),
    name="cost_fn",
)
objective.add(cost_fn)
optimizer = th.LevenbergMarquardt(objective, max_iteration=15, step_size=1.0)

layer = th.TheseusLayer(optimizer=optimizer)

with torch.no_grad():
    output, info = layer.forward(
        input_tensors={"Th_vec": torch.zeros(6).unsqueeze(0)},  # zeros as initial guess
        optimizer_kwargs={"track_best_solution": True, "verbose": True},
    )

print("best solution: ", info.best_solution)
print("target solution: ", Th_target.log())
print(output)
```

Somehow, the program doesn't find an optimal solution here. I believe, the Jacobians are zero, so maybe the `cost_fn` doesn't propagate the gradient. 

Any idea what goes wrong?

Thank you so much for the help!

**Comment by mhubii** [comment]
note, when in the above `cost_fn`

```python
def cost_fn(optim_vars, aux_vars):
    Th_vec = optim_vars[0]
    p, p_prime = aux_vars
    ### <----------------- replace what follows
    Th = lie.SE3.exp(Th_vec.tensor)
    p_prime_hat = Th @ p.tensor
    ###
    err = p_prime_hat - p_prime.tensor
    return err
```

is replaced with
```python
def cost_fn(optim_vars, aux_vars):
    Th_vec = optim_vars[0]
    p, p_prime = aux_vars
    ### <----------------- replacement
    r = th.SO3.exp_map(Th_vec.tensor[0, 3:].unsqueeze(0))
    t = Th_vec.tensor[0, :3].unsqueeze(0).unsqueeze(-1)
    Th = th.SE3(tensor=torch.cat([r.tensor, t], dim=2))
    p_prime_hat = p.tensor @ Th.tensor[:, :3, :3] + Th.tensor[:, :3, 3]
    p_prime_hat = p_prime_hat.squeeze(0)
    ###
    err = p_prime_hat - p_prime.tensor
    return err
```

everything works as intended.

**Comment by luisenp** [comment]
Thanks for reporting this and really sorry for the very long delay! Looking into this now.

**Comment by luisenp** [comment]
Confirmed there is some issue with the LieTensor `@` operator. While we work on fixing this error, one workaround is using the functional library, as shown below. 

```python
import torchlie.functional as lieF

def cost_fn(optim_vars, aux_vars):
    Th_vec = optim_vars[0]
    p, p_prime = aux_vars
    Th = lieF.SE3.exp(Th_vec.tensor)
    p_prime_hat = lieF.SE3.transform(Th, p.tensor)
    err = p_prime_hat - p_prime.tensor
    return err

**Comment by luisenp** [comment]
@mhubii Turns out the tensors created by `LieTensor.exp()` were being created outside the compute graph. Your code should work after #631 is merged. Thanks for reporting!

**Comment by mhubii** [comment]
Thanks for the response @luisenp . No problem , we found a solution that worked. Thank you for sharing this great library.
